openresty-build-tools/kong-ngx-build

```diff
@@ -380,56 +417,88 @@ main() {
-            "-DPURIFY"
-            "no-threads"
-            "--prefix=$OPENSSL_PREFIX"
-            "--openssldir=$OPENSSL_PREFIX"
-          )
-
-          if version_gte $OPENSSL_VER 1.1.0; then
-            OPENSSL_OPTS+=('no-unit-test')
-
-          else
-            OPENSSL_OPTS+=('no-tests')
-          fi
+  if [ -n "$OPENSSL_VER" ]; then
+    if [ ! -f $OPENSSL_INSTALL/bin/openssl ]; then
+      notice "Building OpenSSL..."
+
+      pushd $OPENSSL_DOWNLOAD
+        if (version_lte $OPENSSL_VER 1.0 && [[ ! -d include/openssl ]]) || [[ ! -f Makefile ]]; then
+            OPENSSL_OPTS=(
+              "-g"
+              "shared"
+              "-DPURIFY"
+              "no-threads"
+              "--prefix=$OPENSSL_PREFIX"
+              "--openssldir=$OPENSSL_PREFIX"
+            )
+
+            if version_gte $OPENSSL_VER 1.1.0; then
+              OPENSSL_OPTS+=('no-unit-test')
+
+            else
+              OPENSSL_OPTS+=('no-tests')
+            fi
+
+            if ([[ $CC == "clang" ]] && version_gte $OPENSSL_VER 1.1) || [[ $CC != "clang" ]]; then
+              local ld_opts="-Wl,-rpath,'\$(LIBRPATH)'"
+              if [[ $OS != "Darwin" ]]; then
+                ld_opts="$ld_opts,--enable-new-dtags"
+              fi
 
-          if ([[ $CC == "clang" ]] && version_gte $OPENSSL_VER 1.1) || [[ $CC != "clang" ]]; then
-            local ld_opts="-Wl,-rpath,'\$(LIBRPATH)'"
-            if [[ $OS != "Darwin" ]]; then
-              ld_opts="$ld_opts,--enable-new-dtags"
+              OPENSSL_OPTS+=("$ld_opts")
             fi
 
-            OPENSSL_OPTS+=("$ld_opts")
-          fi
+            if [ $DEBUG == 1 ]; then
+              OPENSSL_OPTS+=('-d')
+            fi
 
-          if [ $DEBUG == 1 ]; then
-            OPENSSL_OPTS+=('-d')
-          fi
+          eval ./config ${OPENSSL_OPTS[*]}
+        fi
 
-        eval ./config ${OPENSSL_OPTS[*]}
-      fi
+        if version_gte $OPENSSL_VER 1.1.0; then
+          make -j$NPROC
+        else
+          make
+        fi
 
-      if version_gte $OPENSSL_VER 1.1.0; then
-        make -j$NPROC
-      else
-        make
-      fi
+        make install_sw DESTDIR=${OPENSSL_DESTDIR}
+      popd
 
-      make install_sw DESTDIR=${OPENSSL_DESTDIR}
-    popd
+      succ "OpenSSL $OPENSSL_VER has been built successfully!"
 
-    succ "OpenSSL $OPENSSL_VER has been built successfully!"
+    else
+      succ "OpenSSL $OPENSSL_VER has been built successfully (cached)!"
+    fi
+  fi
 
-  else
-    succ "OpenSSL $OPENSSL_VER has been built successfully (cached)!"
+  if [ -n "$BORINGSSL_VER" ]; then
+    if [ ! -f "$BORINGSSL_DOWNLOAD/.openssl/lib/libcrypto.so" ]; then
+      notice "Building BoringSSL..."
+
+      pushd $BORINGSSL_DOWNLOAD
+
+        mkdir -p build
+        pushd build
+          CC=/usr/bin/clang \
+          CXX=/usr/bin/clang++ \
+          CPP=/usr/bin/clang-cpp \
+          LD=/usr/bin/ld.lld \
+          cmake -GNinja -DBUILD_SHARED_LIBS=1 -DFIPS=1  ..
```

Comment:
for a general purpose boringssl build this should be fine, but if i understand correctly for fips 
we will need to follow the exact steps as performed in the security policy
(https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp3678.pdf section 12.1), that uses specific version
of build chain. actually we can reuse much of that from envoy (https://github.com/envoyproxy/envoy/blob/a12869fa9e9add4301a700978d5489e6a0cc0526/bazel/external/boringssl_fips.genrule_cmd).
Might need to downgrade the c++ headers as well, the fips validated module doesn't compile
happy with most recent libstdc++ headers (https://github.com/fffonion/lua-resty-openssl/blob/master/.github/workflows/tests.yml#L156 could be helpful)

Reply:
yes, full validation requires those steps and verification, but the 2.7 scope isn't FIPS-validation, just to run with the library in FIPS mode.  the exact version can be specified in the `.requirements` file; perhaps the C++ headers version should be there too?

Reply:
>  just to run with the library in FIPS mode.

in that case this feels okay to me.

> the exact version can be specified in the .requirements file;

Yeah locking the versions in .requirements is indeed a good approach for future 👍